Data/HashMap/Base.hs

```diff
@@ -1949,7 +1949,7 @@ updateOrSnocWithKey f k0 v0 ary0 = go k0 v0 ary0 0 (A.length ary0)
             A.write mary n (L k v)
             return mary
         | otherwise = case A.index ary i of
-            (L kx y) | k == kx   -> A.update ary i (L k (f k v y))
+            (L kx y) | k == kx, (# v2 #) <- f k v y -> A.update ary i (L k v2)
```

Comment:
I'm still a bit mystified how the fix works, but this must be the crucial line.

By forcing the unboxed tuple, we can get rid of the reference to the old value `y`, so it can be GC'd. Is that right?

Reply:
This line does the forcing, so it's half the story. Line 760 constructs `\a _ -> (# a #)` which says how far to evaluate (at least as far as that the `_` has disappeared). Together they get rid of the old value, that previously hung out in the `_`, and is now not in the `(# #)` value.

Reply:
Speaking of this line, that case-plus guard can be rewritten

```haskell
        | L kx y <- A.index ary i
        = if
             | k == kx
             , (# v2 #) <- f k v y
             -> A.update ary i (L k v2)
             | otherwise
             -> go k v ary (i+1) n
```

Personally, I think this seems a little easier to follow.

Reply:
I refactored to morally follow that pattern - the nested `if` is not actually needed as the `kx` and `y` bound by `A.index` aren't used (and importantly, _shouldn't_ be used) in the `go` recursive call - so I hoisted it all up a bit.